client/src/providers/User.js

```diff
@@ -0,0 +1,122 @@
+import React, { createContext, useReducer, useEffect, useMemo } from "react";
+import useLocalStorage from "../hooks/useLocalStorage";
+import { gql, useQuery } from "@apollo/client";
+
+const STORE_KEY = "@CashlessUser@1";
+
+const context = {
+  id: null,
+  isAuthenticated: false,
+  secret: {},
+  token: null,
+  profile: {
+    name: null,
+    email: null,
+    description: null,
+  },
+};
+
+const defaultContext = [context, () => {}];
+
+const UserContext = createContext(defaultContext);
+
+function reducer(state, action) {
+  switch (action.type) {
+    case "set_secret": {
+      try {
+        const newState = {
+          ...state,
+          id: action.value.id,
+          secret: action.value,
+          isAuthenticated: true,
+        };
+        return { ...state, ...newState };
+      } catch {
+        return state;
+      }
+    }
+    case "set_profile": {
+      try {
+        return {
+          ...state,
+          profile: {
+            ...state.profile,
+            ...action.value,
+          },
+        };
+      } catch {
+        return state;
+      }
+    }
+    case "set_token": {
+      try {
+        return { ...state, token: action.value };
+      } catch {
+        return state;
+      }
+    }
+
+    default:
+      return state;
+  }
+}
+
+export const PROFILE = gql`
+  query($id: ID!) {
+    profile(id: $id) {
+      id
+      name
+      description
+    }
+  }
+`;
+
+export const UserProvider = ({ children }) => {
+  const [initialState, setState] = useLocalStorage(STORE_KEY, context);
+  const [state, dispatch] = useReducer(reducer, initialState);
+  let [contextValue, memoDispatch] = useMemo(() => [state, dispatch], [
+    state,
+    dispatch,
+  ]);
+  const {
+    id,
+    token,
+    secret,
+    profile: { name, description },
+  } = contextValue;
+  const { fetchMore } = useQuery(PROFILE, {
+    variables: { id },
+  });
+  useEffect(() => {
+    if (token && secret && (!name || !description)) {
+      /* TODO: set to locastorage */
```

Comment:
I see 
`  const [initialState, setState] = useLocalStorage(STORE_KEY, context);`
.. above so does this mean we're storing the critical state in local storage ? Or do we need to do more ? 

We definitely want the behavior that the user essentially 'stays logged in' if they close browser and come back..

Reply:
Local storage persists and is only cleaned if the user cleans up it's browser data.